# House Price Prediction
## Problem Definition
The goal of this project is to build a machine learning model
that predicts house prices based on different characteristics
of a house.
### Machine Learning Type
Supervised Learning
### Problem Type
Regression
### Target Variable
price

In [96]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [97]:
df = pd.read_csv("C:/Users/HP/OneDrive/Desktop/ML_prac/datasets/messy_housing_data.csv")


In [98]:
df

,area_sqft,bedrooms,bathrooms,year_built,price,location
0,2696,4.0,1,1962.0,2068779.98,austin
1,2017,6.0,4,2008.0,1539208.46,San Francisco
2,1592,3.0,3,1991.0,557909.7,San Francisco
3,3358,2.0,4,1914.0,NaN,chicago
4,1624,2.0,5,1935.0,716945.96,New York City
...,...,...,...,...,...,...
415,3674,5.0,5,NaN,2664238.68,l.a.
416,939,3.0,5,1965.0,491056.77,Los Angeles
417,4880,3.0,3,2004.0,3709981.28,boston
418,3279,5.0,5,1935.0,1308383.3,Chicago


In [99]:
#Shape of the data
df.shape

(420, 6)

** What does that mean? **

means we have 420 houses and we have 6 variables/columns.

In [100]:
#The structure of the dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 420 entries, 0 to 419
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   area_sqft   420 non-null    int64  
 1   bedrooms    399 non-null    float64
 2   bathrooms   420 non-null    int64  
 3   year_built  384 non-null    float64
 4   price       370 non-null    object 
 5   location    420 non-null    object 
dtypes: float64(2), int64(2), object(2)
memory usage: 19.8+ KB


In [101]:
#Descriptive statistics
df.describe()

,area_sqft,bedrooms,bathrooms,year_built
count,420.000000,399.000000,420.00000,384.000000
mean,2621.888095,3.593985,3.02381,1960.197917
std,1330.825892,2.386218,1.45067,37.715887
min,-100.000000,1.000000,1.00000,1900.000000
25%,1462.500000,2.000000,2.00000,1926.000000
50%,2602.000000,3.000000,3.00000,1959.000000
75%,3709.000000,5.000000,4.00000,1994.000000
max,4999.000000,20.000000,5.00000,2026.000000


In [102]:
#Checking for missing values
df.isnull().sum()

area_sqft      0
bedrooms      21
bathrooms      0
year_built    36
price         50
location       0
dtype: int64

In [103]:
#Check duplicates
df.duplicated().sum()

np.int64(20)

In [104]:
#Show the duplicated rows
df[df.duplicated()]

,area_sqft,bedrooms,bathrooms,year_built,price,location
400,3857,1.0,4,1970.0,1121087.17,los angeles
401,4695,3.0,1,2019.0,1960710.64,Boston
402,4360,6.0,3,1950.0,2551135.51,New York city
403,650,3.0,2,NaN,136580.26,New York City
404,1974,1.0,5,NaN,1147299.91,New York city
405,706,4.0,4,1919.0,NaN,austin
406,2266,1.0,3,1917.0,850749.66,Chicago
407,4281,4.0,4,1976.0,1871382.95,Bostan
408,976,NaN,2,2004.0,287451.49,los angeles
409,2256,3.0,4,1937.0,1763039.92,sf


In [105]:
# Shows both the original rows and their duplicate copies, making it easier to understand what's happening.
df[df.duplicated(keep=False)].sort_values(by=["area_sqft", "bedrooms"])

,area_sqft,bedrooms,bathrooms,year_built,price,location
210,650,3.0,2,NaN,136580.26,New York City
403,650,3.0,2,NaN,136580.26,New York City
84,706,4.0,4,1919.0,NaN,austin
405,706,4.0,4,1919.0,NaN,austin
278,939,3.0,5,1965.0,491056.77,Los Angeles
416,939,3.0,5,1965.0,491056.77,Los Angeles
266,976,NaN,2,2004.0,287451.49,los angeles
408,976,NaN,2,2004.0,287451.49,los angeles
385,1139,2.0,3,1915.0,496344.73,nyc
419,1139,2.0,3,1915.0,496344.73,nyc


In [106]:
#Removing duplicates
df = df.drop_duplicates() 

In [107]:
df.shape

(400, 6)

In [108]:
df.duplicated().sum()

np.int64(0)

In [109]:
#Investigating the price column
df['price'].head(20)

0     2068779.98
1     1539208.46
2       557909.7
3            NaN
4      716945.96
5      574438.17
6            NaN
7            NaN
8      1974364.8
9     1489205.48
10    3830378.09
11     542187.73
12    1461241.11
13           NaN
14    1165755.41
15    1637513.07
16     2965470.5
17     612815.74
18    2470823.55
19     446664.75
Name: price, dtype: object

In [110]:
df['price'].dtype

dtype('O')

In [111]:
df['price'].unique()

array(['2068779.98', '1539208.46', '557909.7', nan, '716945.96',
       '574438.17', '1974364.8', '1489205.48', '3830378.09', '542187.73',
       '1461241.11', '1165755.41', '1637513.07', '2965470.5', '612815.74',
       '2470823.55', '446664.75', '2499620.89', '25710', '1082617.93',
       '329772.59', '693954.16', '1253145.14', '524115.14', '2023729.99',
       '1297552.1', '363624.94', '814621.64', '2551135.51', '1769596.1',
       '659706.61', '526823.25', '1053573.57', '1175350.02', '1693434.38',
       '2664238.68', '170067.87', '1409945.42', '2407961.69',
       '1593555.06', '1741343.06', '1061991.43', '349256.54',
       '2884846.24', '790682.37', '575126.88', '1056683.16', '1796102.58',
       '605836.19', '2615033.93', 'unavailable', '1595071.35',
       '2279975.0', '568549.58', '541651.91', '2084710.67', '528502.71',
       '438478.83', '1353744.45', '39240', '451972.47', '991381.15',
       '2023776.94', '1060132.81', '937610.47', '448012.63', '2440843.48',
       '413490

In [112]:
df['price'].value_counts(dropna=False).head(10)

price
NaN            49
unavailable     7
2068779.98      1
1539208.46      1
557909.7        1
716945.96       1
574438.17       1
1974364.8       1
1489205.48      1
3830378.09      1
Name: count, dtype: int64

In [113]:
(df['price'] == 'unavailable').sum()

np.int64(7)

In [114]:
df['price'] = pd.to_numeric(df['price'], errors = "coerce")

C:\Users\HP\AppData\Local\Temp\ipykernel_19092\3767289678.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price'] = pd.to_numeric(df['price'], errors = "coerce")


In [115]:
df['price'].dtype

dtype('float64')

In [116]:
#Checking for missing values again
df.isnull().sum()

area_sqft      0
bedrooms      20
bathrooms      0
year_built    32
price         56
location       0
dtype: int64

In [117]:
df[df['area_sqft'] <=0]

,area_sqft,bedrooms,bathrooms,year_built,price,location
73,-100,3.0,1,1977.0,1060132.81,san francisco
128,-100,2.0,3,1904.0,586529.68,nyc
190,-100,1.0,3,1991.0,2086097.89,San Francisco
384,-100,2.0,3,2006.0,1454516.52,San Francisco


In [118]:
df['area_sqft'].describe()

count     400.000000
mean     2620.532500
std      1329.759422
min      -100.000000
25%      1462.500000
50%      2628.000000
75%      3705.750000
max      4999.000000
Name: area_sqft, dtype: float64

In [119]:
df.loc[df["area_sqft"] <= 0, "area_sqft"] = np.nan

In [120]:
df["area_sqft"].isnull().sum()

np.int64(4)

In [121]:
df["location"].value_counts()

location
san francisco     28
sf                26
miami             22
Austin            22
austin            20
Chicago           20
LA                19
San Francisco     19
New York city     18
l.a.              17
New York City     17
nyc               17
chicago           16
Los Angeles       16
Boston            16
boston            15
NYC               11
los angeles        9
Miami              7
Austin             6
San Francisco      5
Chicgo             5
Los Angeles        5
Bostan             5
Boston             4
New York City      4
CHICAGO            3
Chicago            3
Miami              3
Sf                 3
AUSTIN             3
L.A.               2
BOSTON             2
L.A.               2
NEW YORK CITY      2
AUSTIN             1
LOS ANGELES        1
SF                 1
La                 1
SAN FRANCISCO      1
CHICAGO            1
Nyc                1
MIAMI              1
Name: count, dtype: int64

In [122]:
df["location"]=df["location"].str.strip()

C:\Users\HP\AppData\Local\Temp\ipykernel_19092\3015142855.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["location"]=df["location"].str.strip()


In [123]:
df["location"]=df["location"].str.lower()

C:\Users\HP\AppData\Local\Temp\ipykernel_19092\3526628287.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["location"]=df["location"].str.lower()


In [124]:
df["location"].value_counts()

location
san francisco    53
austin           52
chicago          43
new york city    41
boston           37
miami            33
los angeles      31
sf               30
nyc              29
l.a.             21
la               20
chicgo            5
bostan            5
Name: count, dtype: int64

In [125]:
location_mapping = {
    'nyc':'new york city', 'l.a.':'los angeles', 'sf':'san francisco',
    'la':'los angeles', 'chicgo':'chicago','bostan':'boston',
    'new york city':'new york city', 'san francisco':'san francisco', 
    'miami':"miami"
}

In [126]:
df['location']=df['location'].replace(location_mapping)

C:\Users\HP\AppData\Local\Temp\ipykernel_19092\4083835098.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['location']=df['location'].replace(location_mapping)


In [127]:
df['location'].value_counts()

location
san francisco    83
los angeles      72
new york city    70
austin           52
chicago          48
boston           42
miami            33
Name: count, dtype: int64

In [128]:
df['location'].unique()

array(['austin', 'san francisco', 'chicago', 'new york city', 'miami',
       'los angeles', 'boston'], dtype=object)

In [129]:
df.isnull().sum()

area_sqft      4
bedrooms      20
bathrooms      0
year_built    32
price         56
location       0
dtype: int64

In [130]:
df.isnull().mean()*100

area_sqft      1.0
bedrooms       5.0
bathrooms      0.0
year_built     8.0
price         14.0
location       0.0
dtype: float64

In [131]:
df[["area_sqft","bedrooms","bathrooms","year_built","price"]].describe()

,area_sqft,bedrooms,bathrooms,year_built,price
count,396.000000,380.000000,400.000000,368.000000,3.440000e+02
mean,2648.012626,3.618421,3.017500,1960.228261,1.371330e+06
std,1307.845418,2.413502,1.455159,37.668887,9.969063e+05
min,515.000000,1.000000,1.000000,1900.000000,9.640000e+03
25%,1527.000000,2.000000,2.000000,1926.750000,6.535463e+05
50%,2653.500000,3.000000,3.000000,1959.000000,1.179630e+06
75%,3709.000000,5.000000,4.000000,1994.000000,1.878623e+06
max,4999.000000,20.000000,5.000000,2026.000000,8.520000e+06


In [132]:
df[["area_sqft","bedrooms","bathrooms","year_built","price"]].median()

area_sqft        2653.50
bedrooms            3.00
bathrooms           3.00
year_built       1959.00
price         1179629.52
dtype: float64

In [133]:
df[["area_sqft","bedrooms","bathrooms","year_built","price"]].mean()

area_sqft     2.648013e+03
bedrooms      3.618421e+00
bathrooms     3.017500e+00
year_built    1.960228e+03
price         1.371330e+06
dtype: float64